subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -475,65 +466,86 @@ public void useFailureHandler(TaskFailureHandler handler) {
         this.failureHandler = handler;
     }
 
-    public TaskInfo getTaskToExecute() {
-        lock.lock();
-        try {
-            while (true) {
+    @Override
+    public boolean executeWithTask(WorkerLease parentWorkerLease, final Action<TaskInfo> taskExecution) {
+        final AtomicReference<TaskInfo> selected = new AtomicReference<TaskInfo>();
+        final AtomicBoolean canExecute = new AtomicBoolean();
+        final ResourceLock workerLease = parentWorkerLease.createChild();
+        coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
+            @Override
+            public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
                 if (cancellationToken.isCancellationRequested()) {
                     if (abortExecution()) {
                         tasksCancelled = true;
                     }
                 }
-                TaskInfo nextMatching = null;
-                boolean allTasksComplete = true;
-                Iterator<TaskInfo> iterator = executionQueue.iterator();
+
+                final Iterator<TaskInfo> iterator = executionQueue.iterator();
                 while (iterator.hasNext()) {
-                    TaskInfo taskInfo = iterator.next();
-                    allTasksComplete = allTasksComplete && taskInfo.isComplete();
-                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete() && canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
-                        nextMatching = taskInfo;
-                        iterator.remove();
-                        break;
+                    final TaskInfo taskInfo = iterator.next();
+                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete()) {
+                        coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
```

Comment:
Yeah, that's why it's there.  `reset()` seems like a reasonable solution.

I've captured this in a follow up story.